### ISSUE TITLE
Screenreader Accessibility: Allow navigating to the message list using the Tab key, andselecting messages using the arrow keys 

### ISSUE THREAD

**Issue by robin24** [josh-seagrave/EasyCPDLC#37]
Hey Josh,

I'm a member of the vibrant blind flight simming community that has started to develop in recent years, most of us fly PMDG aircraft using Prepar3D as they have recently been made accessible using the Talking Flight Monitor application, which can be found at talkingflightmonitor.com.

Many of us (myself included) are very active on VATSIM even participating in major events such as CTP, and so we're also constantly following along with new developments in the online ATC world.

I therefore decided a few days back to give EasyCPDLC a try to see how it all works, and was happy to find that the program is mostly accessible using the NVDA screenreader - kudos to you for that!!!

However, one big issue that I found which makes this impossible for me and our community to use for the time being is that it does not appear to be possible to read received messages, as NVDA is unable to focus them.

The issue here is that as of right now, it looks like keyboard focus cannot be moved to the message list, making this a keyboard-focusable UI object would most likely solve this problem.

A good example of a program where this works really well is vPilot, where the area that displays text messages received on VATSIM in a read-only text field that is fully accessible.

I know you're probably really busy working on EasyCPDLC*s core features and enhancements that will benefit your wider user base, but was wondering if you could perhaps look into the possibility of making this list more accessible when you have a moment?

I'm not super familiar with programming on Windows I'm afraid, but my hope is that this would only require a couple property changes on this UI element to make it work.

I'd really love to jump into this new world of CPDLC communication on some of my flights, so would really appreciate this!

Here are some additional links that you might find useful / interesting:
- Our BVIPilots website which contains more info on how we do things as blind / visually impaired simmers: https://bvipilots.net
- - The NVDA screenreader, which is free and open-source software, can be located here: https://nvaccess.org
Thanks for possibly considering this, please always feel free to reach out to me with any questions!

Robin

**Comment by josh-seagrave** [comment]
Hi Robin,

Thankyou so much for bringing this to my attention. From the ground up I had intended EasyCPDLC to be accessible by all, so it was by no means my intention to omit features that would allow the blind and visually impaired community to use my software. Please accept my sincerest apologies.

I will investigate this afternoon and make remedying this a priority. I can't guarantee that I will be able to fix it and release a new build in time for CTP unfortunately, but I will try my best. I suspect you are right in that I never made the message list keyboard-accessible, which _should_ be a relatively straightforward fix. I'll take the opportunity to check that everything else is keyboard accessible while I'm at it, so hopefully this won't be an issue going forwards.

Kind Regards,

Joshua

**Comment by robin24** [comment]
Hey Joshua,

Wow, thank you so much for your super kind response!

I really really appreciate your wonderful intentions of making this fully accessible from the start, but realize this can be tricky when you don’t have users relying on assistive technology like a screenreader to provide feedback.

So, honestly don’t feel bad about it at all, it happens and I’m super excited to hear that you’ll be looking into fixing it!

If you come up with a potential fix and would like for me to take a look before releasing an updated version to all users, you’re more than welcome to reach out with a patch and I’ll give it a spin as soon as I can.

Also, having this ready to rock for tomorrow’s CTP would definitely be suuuuuper freaking awesome (especially since it would provide plenty of opportunity for testing), but I realize this might take some code refactoring etc so seriously do take your time with it, even just knowing that you’re interested in fixing this has me really excited already!

Cheers,
Robin

**Comment by josh-seagrave** [comment]
Hi Robin,

Looking at my code now, I understand what the root problem is. To clean up the UI and assist with the styling of elements, EasyCPDLC uses labels in most cases in lieu of textboxes. By default, and unbeknownst to me at the time, labels are not focusable by tabbing, and require some additional properties to be set. I'm currently working my way through to ensure that everything that requires focus can be focussed.

Would you mind answering a question for me? With NVDA and the various accessibility features you use, do you prefer UI elements to be accessible by keyboard or mouse functions? The reason I ask is that I might make some additional inputs possible to allow messages to be opened by keyboard only, if that would be more convenient.

Let me know what you think!

Josh

**Comment by robin24** [comment]
Hey Josh,

Wow, that sounds great!

Keyboard navigation is by far the preferable means of navigating for screenreader users, and in most cases the keyboard is used exclusively for all tasks.

NVDA and other screenreaders do have functions to simulate moving the mouse pointer and clicking items on the screen, but this can be unreliable especially when there are other programs taking up the majority of the screen (such as a sim window).

So, if you could make it possible to navigate and respond to messages using the keyboard, that would be huge! :-)

Robin

**Comment by josh-seagrave** [comment]
Thanks for your input, I'll get to work!

Josh

**Comment by josh-seagrave** [comment]
Hi Robin,

I've built a development version for you to test, if you'd be interested. The updated .exe is here: https://drive.google.com/file/d/1ENPiheoijziQ6DmEHuHhK-zKdt1ILtwI/view?usp=sharing
All you should have to do is replace the normal .exe (maybe make a backup first) with this one.

I've made a number of changes that hopefully will make your experience a lot more manageable, I'll list them briefly here:

- Message logs should now be accessible to you by keyboard. I've made the tab logic as sensible as I can, at least for me, so hopefully it's relatively straightforward. Pressing space or enter on a message will take you to the message window where you should be able to interact with and reply to messages that are received, and the message response options all respond to the space key or the enter key.
- The TELEX and ATC menus should now be fully operable through just the keyboard, using tab to iterate through menus and data fields. All buttons should respond to the space bar or enter key.
- The settings menu is now accessible by tab, where it wasn't before.
- I wasn't sure whether your screen reader needed labels and timestamps to be accessible or not, so I've made them accessible just in case. If this is not required or more of a help than a hinderance I can remove it.

Please let me know how you get on with this version, if you have the time or inclination to test it. I'd love your feedback for how I can make the layout more friendly or add/remove features to make the program easier to use.

Apologies that it's taken me a few hours to get this version sorted, but hopefully the improvements I've made are worth the wait.

Kind Regards,

Joshua

**Comment by robin24** [comment]
Hey Josh,

Thank you so much for this truly amazing work!!! I'm setting up for a flight right now and hoping to maybe get some CPDLC service along the way, I'll let you know how it went as soon as it's done! :-)

Robin
Am Freitag, April 01, 2022 19:58 CEST, schrieb Joshua Seagrave ***@***.***>:
 
 
      Hi Robin,
I've built a development version for you to test, if you'd be interested. The updated .exe is here: https://drive.google.com/file/d/1ENPiheoijziQ6DmEHuHhK-zKdt1ILtwI/view?usp=sharing
All you should have to do is replace the normal .exe (maybe make a backup first) with this one.
I've made a number of changes that hopefully will make your experience a lot more manageable, I'll list them briefly here:
 * Message logs should now be accessible to you by keyboard. I've made the tab logic as sensible as I can, at least for me, so hopefully it's relatively straightforward. Pressing space or enter on a message will take you to the message window where you should be able to interact with and reply to messages that are received, and the message response options all respond to the space key or the enter key. * The TELEX and ATC menus should now be fully operable through just the keyboard, using tab to iterate through menus and data fields. All buttons should respond to the space bar or enter key. * The settings menu is now accessible by tab, where it wasn't before. * I wasn't sure whether your screen reader needed labels and timestamps to be accessible or not, so I've made them accessible just in case. If this is not required or more of a help than a hinderance I can remove it.
Please let me know how you get on with this version, if you have the time or inclination to test it. I'd love your feedback for how I can make the layout more friendly or add/remove features to make the program easier to use.
Apologies that it's taken me a few hours to get this version sorted, but hopefully the improvements I've made are worth the wait.
Kind Regards,
Joshua

**Comment by josh-seagrave** [comment]
I'm glad we're making progress!

Yes, I'll look into that right away, thanks for letting me know

**Comment by josh-seagrave** [comment]
Found the cause of that, forgot to update those two functions to allow keyboard inputs, they're now fixed :)

**Comment by robin24** [comment]
Holy smokes, you're a hero!!! Is that updated build available through the link you posted earlier?

**Comment by josh-seagrave** [comment]
> Holy smokes, you're a hero!!! Is that updated build available through the link you posted earlier?

yep, should be :)

**Comment by robin24** [comment]
Downloaded and just tested this out with the ESOS controller who was really helpful in helping me test this out, so I am happy to let you know this is indeed fully usable now!!!

There definitely are some things here and there that could be improved - for instance, right now it's not possible to move through the text of messages by letter with the left and right arrow keys or to select text which would be super useful to have, however those aren't at all critical for the time being.

Super excited to try this out tomorrow for my CTP flight and I'll be sure to get back to you with some more detailed feedback, but the progress you've made today is truly fabulous!

Thank you so much again for all this amazing work and I hope you have a fun day controlling tomorrow!

Robin

**Comment by josh-seagrave** [comment]
Thankyou so much for your help and the suggestions! I'll look into making the message text selectable, might require a bit of work to figure out exactly how to do that but I'll add it to the list. Might have to use textboxes instead of labels for that, but at least it's largely functional now.

Enjoy tomorrow! I'll keep an eye out for you :)

**Comment by robin24** [comment]
Hey Josh,

Hope you're doing great!

It was a pleasure speaking to you on frequency a few weeks back when some fellow BVI pilots and I were doing a group flight.
Congrats on the new release of EasyCPDLC Beta 0.2.0.0, I just downloaded it to give this a spin.
Unfortunately, I discovered that somehow messages are no longer keyboard focusable in this build, kind of how it was before you made accessibility fixes in the old release.
Just wanted to give you a heads-up about this so that perhaps you could kindly take a look at why this has regressed when you have the time, not urgent at all for now as I and others can continue to use the old version, which has been working great during all this time BTW.

Absolutely feel free to reach out if you have any more questions on the issue I'm having, and of course I'll be happy to test things out once a fix is available.

Have a great evening!

**Comment by quassbutreally** [comment]
> Hope you're doing great!

Hi Robin,

Thanks for letting me know about this. I'll have a look and get a fix implemented ASAP
